# 03 - Interactive Viz

## Deadline

Wednesday November 8th, 2017 at 11:59PM

## Important Notes

- Make sure you push on GitHub your Notebook with all the cells already evaluated
- Note that maps do not render in a standard Github environment : you should export them to HTML and link them in your notebook.
- Remember that `.csv` is not the only data format. Though they might require additional processing, some formats provide better encoding support.
- Don't forget to add a textual description of your thought process, the assumptions you made, and the solution you plan to implement!
- Please write all your comments in English, and use meaningful variable names in your code

## Background

In this homework we will be exploring interactive visualization, which is a key ingredient of many successful data visualizations (especially when it comes to infographics).

Unemployment rates are major economic metrics and a matter of concern for governments around the world. Though its definition may seem straightforward at first glance (usually defined as the number of unemployed people divided by the active population), it can be tricky to define consistently. For example, one must define what exactly unemployed means : looking for a job ? Having declared their unemployment ? Currently without a job ? Should students or recent graduates be included ? We could also wonder what the active population is : everyone in an age category (e.g. `16-64`) ? Anyone interested by finding a job ? Though these questions may seem subtle, they can have a large impact on the interpretation of the results : `3%` unemployment doesn't mean much if we don't know who is included in this percentage. 

In this homework you will be dealing with two different datasets from the statistics offices of the European commission ([eurostat](http://ec.europa.eu/eurostat/data/database)) and the Swiss Confederation ([amstat](https://www.amstat.ch)). They provide a variety of datasets with plenty of information on many different statistics and demographics at their respective scales. Unfortunately, as is often the case is data analysis, these websites are not always straightforward to navigate. They may include a lot of obscure categories, not always be translated into your native language, have strange link structures, … Navigating this complexity is part of a data scientists' job : you will have to use a few tricks to get the right data for this homework.

For the visualization part, install [Folium](https://github.com/python-visualization/folium) (*HINT*: it is not available in your standard Anaconda environment, therefore search on the Web how to install it easily!). Folium's `README` comes with very clear examples, and links to their own iPython Notebooks -- make good use of this information. For your own convenience, in this same directory you can already find two `.topojson` files, containing the geo-coordinates of 

- European countries (*liberal definition of EU*) (`topojson/europe.topojson.json`, [source](https://github.com/leakyMirror/map-of-europe))
- Swiss cantons (`topojson/ch-cantons.topojson.json`) 

These will be used as an overlay on the Folium maps.


In [2]:
import os
import pandas as pd
import json
import folium
import re
import numpy as np

pd.options.mode.chained_assignment = None  # default='warn'

# Assignment
## Question 1
 Go to the [eurostat](http://ec.europa.eu/eurostat/data/database) website and try to find a dataset that includes the european unemployment rates at a recent date.

   Use this data to build a [Choropleth map](https://en.wikipedia.org/wiki/Choropleth_map) which shows the unemployment rate in Europe at a country level. Think about [the colors you use](https://carto.com/academy/courses/intermediate-design/choose-colors-1/), how you decided to [split the intervals into data classes](http://gisgeography.com/choropleth-maps-data-classification/) or which interactions you could add in order to make the visualization intuitive and expressive. Compare Switzerland's unemployment rate to that of the rest of Europe.



## Choice of the dataset
lfsq_urgan 


In [3]:
rate_filepath='./dataEurope/lfsa_urgan.tsv'
#number_filepath='./dataEurope/lfsa_ugan.tsv'

def import_file(filepath):
    df2=pd.read_table(filepath)
    s=df2[df2.columns[0]]
    df1=s.str.split(pat=',', expand=True)
    df1.columns=('Unit', 'Sex', 'Age', 'Citizen', 'Country')
    df2.drop(df2.columns[0],axis=1,inplace=True)
    full_df=pd.concat([df1,df2], axis=1)
    #Cleaning the series for the most recent entry (2016)
    s_lastyear=full_df[full_df.columns[5]]
    for idx in s_lastyear.index:
        non_decimal = re.compile(r'[^\d.]+')
        s_lastyear[idx]=non_decimal.sub('',s_lastyear[idx])
        try:
            s_lastyear[idx]=float(s_lastyear[idx])
        except:
            s_lastyear[idx]=np.nan
    return pd.concat([full_df[['Unit', 'Sex', 'Age', 'Citizen', 'Country']],s_lastyear],axis=1)

df_rate=import_file(rate_filepath)
df_rate.rename(columns={df_rate.columns[5]:'Rate'}, inplace=True)
#df_rate.Country

In [4]:
c=df_rate.Country.unique()
c.size

39

In [5]:
#df_both=pd.concat([df_rate,df_numb[df_numb.columns[5]]],axis=1)
cropped1=df_rate[df_rate.Age=='Y15-74']
cropped2=cropped1[cropped1.Sex=='T']
cropped3=cropped2[cropped2.Citizen=='TOTAL']
cropped3['Rate']
test=pd.DataFrame({'Country': ['FR', 'CH'], 'Rate':[18,2]})
c=cropped3.Country
c.values.size

39

In [34]:
m=folium.Map(location=[22.3, 6.66])
path_EU='./topojson/europe.topojson.json'
eu_json_data=json.load(open(path_EU))
eu_map=folium.Map(location=[46, 7],zoom_start=3)
eu_map.choropleth(geo_data=eu_json_data,topojson='objects.europe', data=cropped3, columns=['Country', 'Rate'],
                 key_on='feature.id', fill_color='YlGn', fill_opacity=0.8, line_opacity=0.6,
             legend_name='Unemployement rate(%)')
#eu_map


## Question 2
Go to the [amstat](https://www.amstat.ch) website to find a dataset that includes the unemployment rates in Switzerland at a recent date.

   > *HINT* Go to the `details` tab to find the raw data you need. If you do not speak French, German or Italian, think of using free translation services to navigate your way through. 

   Use this data to build another Choropleth map, this time showing the unemployment rate at the level of swiss cantons. Again, try to make the map as expressive as possible, and comment on the trends you observe.

   The Swiss Confederation defines the rates you have just plotted as the number of people looking for a job divided by the size of the active population (scaled by 100). This is surely a valid choice, but as we discussed one could argue for a different categorization.

   Copy the map you have just created, but this time don't count in your statistics people who already have a job and are looking for a new one. How do your observations change ? You can repeat this with different choices of categories to see how selecting different metrics can lead to different interpretations of the same data.

In [87]:
path='./dataCH/TauxDeChomageCH.csv'
df=pd.read_csv(path, thousands='\'',encoding='utf_16le', 
index_col=['Canton'])

df.drop('Grande région',axis=1,inplace=True)
df.drop('Mois', axis=1, inplace=True)
df.rename(columns={df.columns[1]:'Grande région'}, inplace=True)
df

Région linguistique      Grande région  \
Canton                                                                      
                                                                      NaN   
Berne                                Suisse alémanique  Espace Mittelland   
Soleure                              Suisse alémanique  Espace Mittelland   
Bâle-Ville                           Suisse alémanique    Nordwestschweiz   
Bâle-Campagne                        Suisse alémanique    Nordwestschweiz   
Argovie                              Suisse alémanique    Nordwestschweiz   
Zurich                               Suisse alémanique             Zürich   
Glaris                               Suisse alémanique         Ostschweiz   
Schaffhouse                          Suisse alémanique         Ostschweiz   
Appenzell Rhodes-Extérieures         Suisse alémanique         Ostschweiz   
Appenzell Rhodes-Intérieures         Suisse alémanique         Ostschweiz   
St-Gall                              Suisse alémanique         Ostschweiz   
Grisons                              Suisse alémanique         Ostschweiz   
Thurgovie                            Suisse alémanique         Ostschweiz   
Lucerne                              Suisse alémanique     Zentralschweiz   
Uri                                  Suisse alémanique     Zentralschweiz   
Schwyz                               Suisse alémanique     Zentralschweiz   
Obwald                               Suisse alémanique     Zentralschweiz   
Nidwald                              Suisse alémanique     Zentralschweiz   
Zoug                                 Suisse alémanique     Zentralschweiz   
Vaud                          Suisse romande et Tessin   Région lémanique   
Valais                        Suisse romande et Tessin   Région lémanique   
Genève                        Suisse romande et Tessin   Région lémanique   
Fribourg                      Suisse romande et Tessin  Espace Mittelland   
Neuchâtel                     Suisse romande et Tessin  Espace Mittelland   
Jura                          Suisse romande et Tessin  Espace Mittelland   
Tessin                        Suisse romande et Tessin             Ticino   
NaN                                              Total                NaN   

                                 Janvier 2017     Janvier 2017.1  \
Canton                                                             
                              Taux de chômage  Chômeurs inscrits   
Berne                                       3             16'954   
Soleure                                   3.2              4'744   
Bâle-Ville                                4.2              4'181   
Bâle-Campagne                             3.1              4'659   
Argovie                                   3.5             12'622   
Zurich                                    3.9             32'387   
Glaris                                    2.5                567   
Schaffhouse                               3.6              1'587   
Appenzell Rhodes-Extérieures                2                610   
Appenzell Rhodes-Intérieures              1.4                129   
St-Gall                                   2.7              7'597   
Grisons                                   1.9              2'117   
Thurgovie                                 2.6              3'967   
Lucerne                                   2.2              4'985   
Uri                                       1.5                297   
Schwyz                                      2              1'794   
Obwald                                      1                217   
Nidwald                                   1.3                305   
Zoug                                      2.6              1'756   
Vaud                                      5.2             20'672   
Valais                                    5.2              9'059   
Genève                                    5.7             13'306   
Fribourg                

## Question 3
 Use the [amstat](https://www.amstat.ch) website again to find a dataset that includes the unemployment rates in Switzerland at recent date, this time making a distinction between *Swiss* and *foreign* workers.

   The Economic Secretary (SECO) releases [a monthly report](https://www.seco.admin.ch/seco/fr/home/Arbeit/Arbeitslosenversicherung/arbeitslosenzahlen.html) on the state of the employment market. In the latest report (September 2017), it is noted that there is a discrepancy between the unemployment rates for *foreign* (`5.1%`) and *Swiss* (`2.2%`) workers. 

   Show the difference in unemployment rates between the two categories in each canton on a Choropleth map (*hint* The easy way is to show two separate maps, but can you think of something better ?). Where are the differences most visible ? Why do you think that is ?

   Now let's refine the analysis by adding the differences between age groups. As you may have guessed it is nearly impossible to plot so many variables on a map. Make a bar plot, which is a better suited visualization tool for this type of multivariate data.


## Question 4 (Bonus)

 *BONUS*: using the map you have just built, and the geographical information contained in it, could you give a *rough estimate* of the difference in unemployment rates between the areas divided by the [Röstigraben](https://en.wikipedia.org/wiki/R%C3%B6stigraben)?